Throughput sample (#12871)
nrfconnect/sdk-nrf: samples/wifi/throughput/README.rst

.. _wifi_throughput_sample:

Wi-Fi: Throughput
#################

.. contents::
   :local:
   :depth: 2

> The Throughput sample demonstrates how to measure the throughputs for Nordic Semiconductor's Wi-Fi® chipsets.

Reviewer: Let's call this "IP networking throughput", to make it more specific.
Suggestion
"The Throughput sample provides a framework for users to measure the achievable IP networking throughput of applications using Nordic Semiconductor's Wi-Fi® chipsets."

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

Overview
********

> The sample demonstrates how to measure network performance using **zperf** tool in different profiles.

Reviewer 2: ```suggestion
The sample demonstrates how to measure network performance using the **zperf** tool in different profiles.
```

> The sample demonstrates how to measure network performance using **zperf** tool in different profiles.

Reviewer: Can we add a link to the ZPERF tool in Zephyr Documentation?

Reviewer: also: network throughput performance? or simply network throughput?

Reviewer: also, what are the different "profiles"? I'd like a more detailed description here, e.g. "different Wi-Fi stack configuration profiles"

Reviewer: Proposal " The sample demonstrates how to measure the network throughput of a Nordic Wi-Fi enabled platform using **zperf** tool under the different Wi-Fi stack configuration profiles"

> Using this sample, the development kit can associate with an access point.

Reviewer: What is the real purpose of adding this sentence?
If I understand correctly, the point should be that it can then receive and send networking traffic which allows users to measure its throughput".
But I could also skip this completely.

Building and running
********************

.. |sample path| replace:: :file:`samples/wifi/throughput`

.. include:: /includes/build_and_run_ns.txt

> Currently, the nRF7002 DK + QSPI configuration is supported.

Reviewer: This could be made as a note. Also, make it more descriptive, e.g.

"The sample is supported on the nRF7002 DK with QSPI as the interface between the nRF5340 host and the nRF7002 device."

To build for the nRF7002 DK, use the ``nrf7002dk_nrf5340_cpuapp`` build target.
The following is an example of the CLI command:

.. code-block:: console

   west build -b nrf7002dk_nrf5340_cpuapp

> To build for the nRF7002 DK with different profiles for Station mode, use the following CLI commands:

Reviewer: Can you add a phrase and a link to where the different profiles are defined? That is, on the corresponding NCS guide?

   .. code-block:: console

      wifi status

   If the connection is established, you should see an output similar to the following:

   .. code-block:: console

      Status: successful
      ==================
      State: COMPLETED
      Interface Mode: STATION
      Link Mode: WIFI 4 (802.11n/HT)
      SSID: xxxxx
      BSSID: xx:xx:xx:xx:xx:xx
      Band: 2.4GHz
      Channel: 11
      Security: WPA2-PSK
      MFP: Optional
      RSSI: -28
      Beacon Interval: 100
      DTIM: 1
      TWT: Not supported

#. Initiate a ping and verify data connectivity using the following commands:

   .. code-block:: console

      net ping <resolved hostname>

   See the following example:

   .. code-block:: console

      net ping 192.168.1.1
      PING 192.168.1.1
      28 bytes from 192.168.1.1 to 192.168.1.1: icmp_seq=1 ttl=64 time=0 ms
      28 bytes from 192.168.1.1 to 192.168.1.1: icmp_seq=2 ttl=64 time=0 ms
      28 bytes from 192.168.1.1 to 192.168.1.1: icmp_seq=3 ttl=64 time=0 ms

#. Run the ``iperf`` commands (TX) on the peer device and the ``zperf`` commands (RX) on the sample as shown in the following table.

>    The Throughput sample test cases require additional software, such as the Wi-Fi **iperf** application.

Reviewer: The last 2 sentences here should be explained before the test steps. It is essential background information of how the sample is supposed to work. So I would move it earlier under the sample description.

Ideally, you should add a **very** simple block diagram, with the nRF7002, the AP, the peer device, etc. It should show what entity run the zperf tool, what entity runs the iperf, what IP protocols are supported, etc.

That part should also explain the configuration. For example, to test TX, you configure zperf on TX and peer on RX, etc.

+------------------------+-----------+-----------+-----------+-----------+
| Profile                | UDP TX    | UDP RX    | TCP TX    | TCP RX    |
|                        |           |           |           |           |
+========================+===========+===========+===========+===========+
| IoT devices            | 5.08 Mbps | 5.43 Mbps | 5.25 Mbps | 3.35Mbps  |
+------------------------+-----------+-----------+-----------+-----------+
| Memory-optimized       | 5.1 Mbps  | 113 Kbps  | 644 Kbps  | 1.86 Mbps |
+------------------------+-----------+-----------+-----------+-----------+
| High performance       | 26.2 Mbps | 13.08 Mbps| 14.1 Mbps | 7.76 Mbps |
+------------------------+-----------+-----------+-----------+-----------+
| TX prioritized         | 26.3 Mbps | 12.20 Mbps| 10.5 Mbps | 4.5 Mbps  |
+------------------------+-----------+-----------+-----------+-----------+
| RX prioritized         | 9.34 Mbps | 13.46 Mbps| 5.48 Mbps | 7.92 Mbps |
+------------------------+-----------+-----------+-----------+-----------+

.. note::
   The measured throughputs, as shown in the table above, are based on tests conducted using the nRF7002 DK.

>    The results represent the best throughput, averaged over three iterations, and were obtained with a good RSSI signal in a clean environment.

Reviewer: This is fine. Ideally you could add a recommendation for how users can achieve a "clean" environment. I.e. what makes the environment less clean? Would it help running this in a chamber, etc.

Author: Hi @ioannisg ,
This setup is placed in RF Chamber. Can I write it this way, 
"The results represent the best throughput, averaged over three iterations, and were obtained with a good RSSI signal in a clean environment (RF Chamber)."

+==============+================================+============================================================+
| UDP          | zperf udp download <dest port> | iperf -c <ipaddress> -i 1 -t <duration> -u -b <baud rate>  |
+--------------+--------------------------------+------------------------------------------------------------+
| TCP          | zperf tcp download <dest port> | iperf -c <ipaddress> -i 1 -t <duration> -b <baud rate>     |
+--------------+--------------------------------+------------------------------------------------------------+

 The sample shows the following output:

   .. code-block:: console

      $ zperf udp download 5001
      UDP server started on port 5001
      <inf> net_zperf: Binding to 192.168.1.13
      <inf> net_zperf: Binding to ::
      <inf> net_zperf: Listening on port 5001
      New session started.
      End of session!
      duration:              10.02 s
      received packets:      10694
      nb packets lost:       33884
      nb packets outorder:   0
      jitter:                        5.20 ms
      rate:                  11.96 Mbps
      $ zperf udp download stop
      UDP server stopped

> #. To stop the ``zperf`` udp/tcp download, use the following command:

Reviewer 3: ```suggestion
#. Stop the ``zperf`` udp/tcp download using the following command:
```

To build for the nRF7002 DK, in memory optimized profile for Station Mode use ``memory-optimized`` overlay configuration.
The following is an example of the CLI command:

.. code-block:: console

   west build -p -b nrf7002dk_nrf5340_cpuapp -- -DOVERLAY_CONFIG=overlay-memory-optimized.conf

To build for the nRF7002 DK, in high performance profile for Station Mode use ``high-performance`` overlay configuration.
The following is an example of the CLI command:

.. code-block:: console

   west build -p -b nrf7002dk_nrf5340_cpuapp -- -DOVERLAY_CONFIG=overlay-high-performance.conf

To build for the nRF7002 DK, in TX prioritized profile for Station Mode use ``tx-prio`` overlay configuration.
The following is an example of the CLI command:

.. code-block:: console

   west build -p -b nrf7002dk_nrf5340_cpuapp -- -DOVERLAY_CONFIG=overlay-tx-prio.conf

To build for the nRF7002 DK, in RX prioritized profile for Station Mode use ``rx-prio`` overlay configuration.
The following is an example of the CLI command:

.. code-block:: console

   west build -p -b nrf7002dk_nrf5340_cpuapp -- -DOVERLAY_CONFIG=overlay-rx-prio.conf

> See also :ref:`cmake_options` for instructions on how to provide CMake options.

Author: Not required. Removed this.

| TCP          | zperf tcp download <dest port> | iperf -c <ipaddress> -i 1 -t <duration> -b <baud rate>     |
+--------------+--------------------------------+------------------------------------------------------------+

 You will see the output similar to the following:

   .. code-block:: console

      $ zperf udp download 5001
      UDP server started on port 5001
      <inf> net_zperf: Binding to 192.168.1.13
      <inf> net_zperf: Binding to ::
      <inf> net_zperf: Listening on port 5001
      New session started.
      End of session!
      duration:              10.02 s
      received packets:      10694
      nb packets lost:       33884
      nb packets outorder:   0
      jitter:                        5.20 ms
      rate:                  11.96 Mbps
      $ zperf udp download stop
      UDP server stopped

#. To stop the zperf udp/tcp download, use the following command:

   .. code-block:: console

      zperf <udp/tcp> download stop

   You will see the output similar to the following:

   .. code-block:: console

      $ zperf udp download stop
      UDP server stopped

Results
=======

> The following tables collect a summary of results obtained when throughput tests are run for different profiles.
> 

Reviewer 3: ```suggestion
The following table collects a summary of results obtained when throughput tests are run for different profiles.
```

+------------------------+-----------+-----------+-----------+-----------+

> | Profile                | UDP Tx    | UDP Rx    | TCP Tx    | TCP Rx    |

Reviewer 3: ```suggestion
| Profile                | UDP TX    | UDP RX    | TCP TX    | TCP RX    |
```

|                        |           |           |           |           |
+========================+===========+===========+===========+===========+
| IoT Devices            | 5.08 Mbps | 5.43 Mbps | 5.25 Mbps | 3.35Mbps  |
+------------------------+-----------+-----------+-----------+-----------+

> | Memory optimized       | 5.1 Mbps  | 113 Kbps  | 644 Kbps  | 1.86 Mbps |

Reviewer 3: ```suggestion
| Memory-optimized       | 5.1 Mbps  | 113 Kbps  | 644 Kbps  | 1.86 Mbps |
```

+------------------------+-----------+-----------+-----------+-----------+
| High performance       | 26.2 Mbps | 13.08 Mbps| 14.1 Mbps | 7.76 Mbps |
+------------------------+-----------+-----------+-----------+-----------+
| TX prioritized         | 26.3 Mbps | 12.20 Mbps| 10.5 Mbps | 4.5 Mbps  |
+------------------------+-----------+-----------+-----------+-----------+
| RX prioritized         | 9.34 Mbps | 13.46 Mbps| 5.48 Mbps | 7.92 Mbps |
+------------------------+-----------+-----------+-----------+-----------+

>    .. note::
> 
>       These tests were run with test setup inside RF Sheildbox and the device is connected to AP in 5 GHz band.
>       Therefore, the results are representative and might change with adjustments in the test setup.

Reviewer 3: ```suggestion
 .. note::

    These tests were run with a test setup inside the RF sheildbox, and the device is connected to the access point in the 5 GHz band.
    Therefore, the results are representative and might change with adjustments in the test setup.
```

.. _wifi_throughput_sample:

Wi-Fi: Throughput
#################

.. contents::
   :local:
   :depth: 2

The Throughput sample provides a framework for users to measure the achievable IP networking throughput of applications using Nordic Semiconductor's Wi-Fi® chipsets.

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

Overview
********

The sample demonstrates how to measure the network throughput of a Nordic Wi-Fi enabled platform using **zperf** sample application under the different Wi-Fi stack configuration profiles.

> The sample require additional software, such as the Wi-Fi **iperf** application. When the sample runs Wi-Fi UDP/TCP throughput in client mode, a peer device runs UDP/TCP throughput in server mode.

Reviewer: require -> requires

> The sample requires additional software, such as the Wi-Fi **iperf** application. When the sample runs Wi-Fi UDP/TCP throughput in client mode, a peer device runs UDP/TCP throughput in server mode.

Reviewer 3: Seems this suggestion got missed to implement. One sentence per line in rst.
```suggestion
The sample requires additional software, such as the Wi-Fi **iperf** application.
When the sample runs Wi-Fi UDP/TCP throughput in client mode, a peer device runs UDP/TCP throughput in server mode.
```

When the sample runs Wi-Fi UDP/TCP throughput in server mode, a peer device runs UDP/TCP throughput in client mode.

See `Operating with a resource constrained host`_ for different Wi-Fi stack configuration profiles.
See `Network Traffic Generator`_ for instructions on how to use **zperf** and **iperf** applications.

Test setup
==========

The following figure shows a reference test setup.

.. figure:: /images/wifi_throughput_sample_test_setup.png

>      :alt: Wi-Fi Throughput Sample test setup

Reviewer 3: ```suggestion
     :alt: Wi-Fi Throughput sample test setup
```

Reviewer 3: This is missed, I think. Same for line 37.

     Wi-Fi Throughput sample test setup

The reference test setup shows the connections between the following devices:

> * nRF7002 DK on which the throughput sample runs **zperf**.

Reviewer 3: ```suggestion
* The nRF7002 DK on which the throughput sample runs the **zperf** application.
```

> * Wi-Fi peer device that runs **iperf**.

Reviewer 3: ```suggestion
* Wi-Fi peer device that runs the **iperf** application.
```

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|
#. Scan for the Wi-Fi networks in range using the following command:

   .. code-block:: console

      wifi scan

   The output should be similar to the following:

   .. code-block:: console

      Scan requested

      Num  | SSID                             (len) | Chan (Band)   | RSSI | Security        | BSSID             | MFP
      1    | abcd                             8     | 11   (2.4GHz) | -18  | WPA2-PSK        | xx:xx:xx:xx:xx:xx | Disable
      2    | efgh                             7     | 40   (5GHz  ) | -21  | OPEN            | xx:xx:xx:xx:xx:xx | Disable
      3    | mnopq                            8     | 6    (2.4GHz) | -29  | OPEN            | xx:xx:xx:xx:xx:xx | Disable
      6    | stuvwx                           24    | 11   (2.4GHz) | -47  | WPA3-SAE        | xx:xx:xx:xx:xx:xx | Required

#. Connect to your preferred network using the following command:

   .. code-block:: console

      wifi connect <SSID> <passphrase>

   ``<SSID>`` is the SSID of the network you want to connect to, and ``<passphrase>`` is its passphrase.

> #. Check the connection status after at least 30 seconds have passed, using the following command:

Reviewer 3: ```suggestion
#. Check the connection status after 30 seconds, using the following command:
```

   .. code-block:: console

      $ zperf udp download 5001
      UDP server started on port 5001
      <inf> net_zperf: Binding to 192.168.1.13
      <inf> net_zperf: Binding to ::
      <inf> net_zperf: Listening on port 5001
      New session started.
      End of session!
      duration:              10.02 s
      received packets:      10694
      nb packets lost:       33884
      nb packets outorder:   0
      jitter:                        5.20 ms
      rate:                  11.96 Mbps

#. Stop the ``zperf`` UDP/TCP download using the following command:

   .. code-block:: console

      zperf <udp/tcp> download stop

   The sample shows the following output:

   .. code-block:: console

      $ zperf udp download stop
      UDP server stopped

Results
=======

The following table collects a summary of results obtained when throughput tests are run for different profiles.

+------------------------+-----------+-----------+-----------+-----------+
| Profile                | UDP TX    | UDP RX    | TCP TX    | TCP RX    |
|                        |           |           |           |           |
+========================+===========+===========+===========+===========+

> | IoT devices            | 5.08 Mbps | 5.43 Mbps | 5.25 Mbps | 3.35Mbps  |

Reviewer 3: ```suggestion
| IoT devices            | 5.08 Mbps | 5.43 Mbps | 5.25 Mbps | 3.35 Mbps  |
```
